pkg/network/ebpf/c/prebuilt/tracer.c

```diff
@@ -1158,52 +1250,300 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
 }
 
+static __always_inline void http_end_response(http_transaction_t *http) {
+    if (http->state != HTTP_RESPONDING) {
+        return;
+    }
+
+    // Retrieve the active batch number for this CPU
+    u32 cpu = bpf_get_smp_processor_id();
+    http_batch_state_t *batch_state = bpf_map_lookup_elem(&http_batch_state, &cpu);
+    if (batch_state == NULL) {
+        log_debug("http batch state not found. should not happen.");
+        return;
+    }
+
+    if (batch_state->pos >= HTTP_BATCH_SIZE) {
+        // We keep incrementing this so we can track how many transactions we're dropping
+        batch_state->pos++;
+        return;
+    }
+
+    http_batch_key_t key;
+    http_prepare_key(cpu, &key, batch_state);
+
+    // Retrieve the batch object
+    http_batch_t *batch = bpf_map_lookup_elem(&http_batches, &key);
+    if (batch == NULL) {
+        log_debug("http batch not found. should not happen. cpu: %d page: %d\n", key.cpu, key.page_num);
+        return;
+    }
+
+    // This redundant information is useful for detecting dirty batch pages on userspace without
+    // incurring on an extra map lookup
+    batch->idx = batch_state->idx;
+
+    // I haven't found a way to avoid this unrolled loop on Kernel 4.4 (newer versions work fine)
+    // If you try to directly write the desired batch slot by doing
+    //
+    //  __builtin_memcpy(&batch->txs[batch_state->pos], http, sizeof(http_transaction_t));
+    //
+    // You get an error like the following:
+    //
+    // R0=inv R1=map_value(ks=4,vs=4816) R2=imm5 R3=imm0 R4=imm0 R6=map_value(ks=48,vs=96) R7=imm1 R8=imm0 R9=inv R10=fp
+    ///809: (79) r2 = *(u64 *)(r6 +88)
+    // 810: (7b) *(u64 *)(r0 +88) = r2
+    // R0 invalid mem access 'inv'
+    //
+    // This is because the value range of the R0 register (holding the memory address of the batch) can't be
+    // figured out by the verifier and thus the memory access can't be considered safe during verification time.
+    // I tried different "tricks" to hint the verifier of this range such as:
+    //
+    // * Ensuring 0 <= batch_slot < HTTP_BATCH_SIZE
+    // * Ensuring that &batch <= &batch->txs[batch_state->pos] <= &batch+1
+    // * Setting HTTP_BATCH_SIZEto a power of 2 and doing &batch->txs[batch_slot&(HTTP_BATCH_PAGES-1)]
+    //
+    // Among other things, but nothing really worked on Kernel 4.4
+    // It seems that indeed support for this type of access by the verifier was added later on:
+    // https://patchwork.ozlabs.org/project/netdev/patch/1475074472-23538-1-git-send-email-jbacik@fb.com/
+    //
+    // What is unfortunate about this is not only that enqueing a HTTP transaction is O(HTTP_BATCH_SIZE),
+    // but also that we can't really increase the batch/page size at the moment because that blows up the eBPF *program* size
+#pragma unroll
+    for (int i = 0; i < HTTP_BATCH_SIZE; i++) {
+        if (i == batch_state->pos) {
+            __builtin_memcpy(&batch->txs[i], http, sizeof(http_transaction_t));
```

Comment:
Did you want to `break` after the `memcpy`?

Reply:
yes, but it doesn't work.
I can't remember now if the problem was with the loop unrolling itself or if the it was something with the verifier